man/rblsmtpd.1
.TH rblsmtpd 1
.SH NAME
rblsmtpd \- block emails from RBL-listed sites and unwanted sources
.SH SYNOPSIS
.B rblsmtpd
.I opts
.I prog
.SH DESCRIPTION
.B rblsmtpd 
works with any SMTP server that can run under
.B tcpserver
or
.BR sslserver
It accepts a series of getopt-style options
.I opts
while
.I prog
consists of one or more arguments.

Normally
.B rblsmtpd
runs
.IR prog .
.I prog
is expected to carry out an SMTP conversation to receive incoming emails.

However,
.B rblsmtpd
does not invoke
.I prog
if it is told to block emails from this client. 
Instead it carries out its own limited SMTP conversation, 
temporarily rejecting all attempts to send an email. 
Meanwhile it prints one line on descriptor 2 to log its activity.

.B rblsmtpd
drops the limited SMTP conversation after 60 seconds, 
even if the client has not quit by then.
.SH "RBL LOOKUP OPTIONS"
.TP
.B \-t \fIn
Change the timeout to
.I n
seconds.
.P
.B Blocked clients
.P
If the 
.B RBLSMTPD 
environment variable is set and is nonempty,
.B rblsmtpd
blocks emails. It uses 
.B RBLSMTPD 
as an error message for the client. Normally
.B rblsmtpd
runs under
.B tcpserver
or better
.B sslserver
and you can use
.BR tcprules
to set 
.B RBLSMTPD 
for selected clients. 

If 
.B RBLSMTPD
is set and is empty,
.B rblsmtpd
does not block emails. 

If 
.B RBLSMTPD 
is not set,
.B rblsmtpd
looks up 
.I TCP6REMOTEIP/TCPREMOTEIP
in the RBL and blocks emails if either
.I TCP6REMOTEIP
or
.I TCPREMOTEIP
is listed here.
.B tcpserver
and 
.B sslserver
set up
.I TCP6REMOTEIP
or
.I TCPREMOTEIP
as the IP address of the remote host.
.TP
.B \-r \fIbase
Use
.I base
as an RBL source. An IPv4 address
.I a.b.c.d
is listed by that source if
.I d.c.b.a.base
has a TXT record.
An IPv6 address is expanded in its inverse nibble format.
.B rblsmtpd
uses the contents of the TXT record as an error message for the client.
.TP
.B \-a \fIbase
Use
.I base
as an anti-RBL source. An IPv4 address
.I a.b.c.d
is anti-listed by that source if
.I d.c.b.a.base
has an A record. In this case
.B rblsmtpd
does not block emails. The same holds for IPv6 addresses (see below) 
and a corresponding AAAA record. 
.P
You may supply any number of
.B \-r
and
.B \-a
options.
.B rblsmtpd
tries each source in turn until it finds one that lists or anti-lists 
according to 
.I TCP6REMOTEIP
or
.IR TCPREMOTEIP .

If you want to run your own RBL source or anti-RBL source for
.BR rblsmtpd ,
you can use
.B rbldns
from the 
.I djbdnscurve6 
package.
.SH "SMTP REPLIES"
Normally, if 
.B RBLSMTPD 
is set,
.B rblsmtpd
uses a 451 error code in it's limited SMTP conversation. 
This tells legitimate clients to try again later. 
It gives innocent relay operators a chance to see the problem, 
prohibit relaying, get off the RBL, and get the email delivered. 

However, if 
.B RBLSMTPD 
begins with a hyphen ('-'),
.B rblsmtpd
removes the hyphen and uses a 553 error code together with the message.
This tells legitimate clients to bounce the email immediately. 

There are several error-handling options for RBL lookups:
.TP
.B \-B
(Default.) Use a 451 error code for IP addresses listed in the RBL.
.TP
.B \-b
Use a 553 error code for IP addresses listed in the RBL.
.TP
.B \-C
(Default.) Handle RBL lookups in a ``fail-open'' mode. 
If an RBL lookup fails temporarily, assume that the address is not listed; 
if an anti-RBL lookup fails temporarily, assume that the address is anti-listed. 
Unfortunately, a knowledgeable attacker can force an RBL lookup or an anti-RBL 
lookup to fail temporarily, so that his email is not blocked.
.TP
.B \-c
Handle RBL lookups in a ``fail-closed'' mode. 
If an RBL lookup fails temporarily, assume that the address is listed 
(but use a 451 error code even with -b). If an anti-RBL lookup fails temporarily, 
assume that the address is not anti-listed (but use a 451 error code even if a 
subsequent RBL lookup succeeds with -b). Unfortunately, this sometimes delays legitimate emails.
.SH "INTERROGATION MODE"
.B rblsmtpd
may be used to only query RBLs and to present the results to
.BR qmail-smtpd
in an interrogation mode.
.TP
.B \-i
interrogation mode; the RBL information is written on descriptor 2 and available
via the environment variable
.B RBLSMTPD
for further decisions.
.SH "GREETDELEY OPTIONS"
Introduce a certain delay in the SMTP connection;
either before or after the RBL, the anti-RBL lookups respectively,
depending whether this option is defined before or after the lookups.
.TP
.B \-w n
provides a delay of 'n' seconds for all connections.
.TP
.B \-W
evaluate the environment variable
.B GREETDELAY
and potentially delay the call to
.BR qmail-smtpd .
.SH "IPv4/IPv6 INVERSE NIBBLE FORMAT"
.B
rblsmtpd 
constructs from the provided IP address 
.I TCP6REMOTEIP/TCPREMOTEIP
the respective inverse nibble format concatenated with the
.I base
name and thus constructing the inverse domain name. Here is a sample for IPv6:
.IP
fe80::1 =>  
1.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.8.e.f.(base)
.P
.B rblsmtpd 
recognizes IPv4-mapped IPv6 addresses 
and strips the IPv6 prefix 
piror to build the IPv4 inverse address
for the RBL lookup.
.SH "SEE ALSO"
sslserver(1),
tcpserver(1),
tcprules(1),
tcprulescheck(1),
tcp-environ(5).


doc/CHANGES
19960311
	version: ucspi-tcp 0.50, alpha.
19960803
	version: ucspi-tcp 0.60, alpha.
19960903
	version: ucspi-tcp 0.70, beta.
19960914
	version: ucspi-tcp 0.71, beta.
19960922
	version: ucspi-tcp 0.72, beta.
19970410
	version: ucspi-tcp 0.73, beta.
19980118
	version: ucspi-tcp 0.80, beta.
19981107
	code: switched to new install system.
	code: eliminated tcpclient -U; moved usage() to default.
	code: switched to strerr in tcpclient.
	code: switched to str_equal in tcpclient.
	portability problem: OpenBSD connect() prohibits 0.0.0.0.
		impact: tcpclient 0 fails. fix: interpret 0 as
		127.0.0.1. hopefully nobody needs 0 to mean primary
		interface.
	code: tcpclient now interprets empty host name as 0.
	doc: added addcr.0, delcr.0.
	doc: added who@.0, date@.0, finger@.0.
	doc: added tcpcat.0, mconnect.0.
	code: added printstatus() in tcpserver.
	code: eliminated numchildren in printpid() in tcpserver.
	code: unblocked SIGCHLD after setting SIGCHLD to SIG_DFL in
		tcpserver, instead of before, to avoid confusion if
		someone sends SIGCHLD manually.
	code: moved printstatus() before fork in tcpserver.
	code: added mconnect-io.
	code: added argv0.
	code: added fixcr.
	doc: revamped BLURB.
	code: added recordio.
	doc: cleaned up tcpclient.0.
	code: added http@.
	code: stopped after 100 bytes in safeappend() in tcpserver.
19981108
	code: added tcpserver -B.
	code: added a few regression tests.
	version: ucspi-tcp 0.83, beta.
19981109
	doc: added http@ to BLURB.
19981110
	code: added exec to mconnect.
	code: tcpclient now ignores port results from getpeername() in
		favor of the port it tried connecting to.
	code: added strport to tcpclient.
	doc: reorganized tcpclient.0.
	doc: reorganized tcpserver.0.
	code: revamped tcpserver messages.
	code: revamped tcpclient messages.
	code: added a few more regression tests.
	code: rewrote recordio for instant output and other features.
	doc: added references in tcpserver.0 and tcpclient.0.
	code: added tcpclient -i option.
	code: added tcprulescheck.
	code: added some tcprules regression tests.
19981111
	code: avoided ndelay in recordio.
	version: ucspi-tcp 0.84, beta.
19981212
	doc: improved description of host 0 in tcpserver.0. tnx to
		several people.
19981218
	doc: reordered sample rules in tcprules.0.
		tnx Harald Hanche-Olsen.
20000120
	internal: revamped packaging.
	internal: massive rewrite. tnx to everyone for bug reports and
		other comments on various versions of tcpclient,
		tcpserver, tcpcontrol, and rblsmtpd: Akihiro Iijima,
		Akihiro Terasaki, Albert J. deVera, Alex Vostrikov,
		Amos Shapira, Anand R. Buddhdev, Andrea Paolini,
		Andrew Pam, Araki Yasuhiro, Arne Wichmann,
		Ayamura Kikuchi, Bart B. B. Hanssens, Bert Gijsbers,
		Bradford M. Shelton, Brendan Kehoe, Brian J. Reichert,
		Brian M. Fisk, Brian O'Reilly, Bruno Wolff,
		Chris Garrigues, Chris Johnson, Christian Wettergren,
		Dale N. Woolridge, Dan M. Vogel, Daniel C. Mahoney,
		Darren Hall, Darren W. Rees, Dave Sill, David J. Walton,
		David P. Smith, David Pool, Dax Kelson, Dirk Jaeckel,
		Dirk Vluegels, Donald E. Blais, Ed Weinberg,
		Edward S. Marshall, Eric A. Perlman, Erik Wallin,
		Faried Nawaz, Fred B. Ringel, Frederik P. Lindberg,
		Gerry Boudreaux, Giles Lean, Grant Holliday,
		Greg D. Patterson, Harald Barth, Harald Fritzsche,
		Harald Hanche-Olsen, Hirofumi Ukawa, Hiroshi Yamashita,
		Hitesh Patel, Ingmar Hupp, J. B. Keith Humphreys,
		Jakub K. Boguslaw, Janos Farkas, Jason R. Mastaler,
		Jeff Hayward, Jeremy Wohl, Jim Littlefield,
		Johan Holmberg, John Bolhuis, John D. Mitchell,
		Jos Backus, Jose Monteiro, Joshua J. Ellis,
		Julie L. Baumler, Karl Lehenbauer, Karsten Thygesen,
		Keith Burdis, Kenny Elliott, Kikuchi Kousuke,
		Kris Kennaway, Lars Balker Rasmussen, Louis S. Theran,
		Lyndon F. Bartels, Mads E. Eilertsen, Mark Delany,
		Martin Mersberger, Matt P. Simerson, Matthew A. Zahorik,
		Michael B. Scher, Michael Handler, Michael Hirohama,
		Michael R. Gile, Michael Salmon, Mikael Suokas,
		Nobuhiro Murata, Patrick M. Kane, Paul R. Rotering,
		Peter Rye, Peter Wilkinson, Petr Novotny,
		Petri Kaukasoina, Raul D. Miller, Richard A. Soderberg,
		Robert W. Luce, Roberto A. Lumbreras, Russ Allbery,
		Russell Nelson, Sean Reifschneider, Shawn A. Clifford,
		Shin Ohira, Shinya O'Hira, Stan Norton,
		Stathy G. Touloumis, Stefan M. Linnemann, Sudish Joseph,
		Thomas E. Erskine, Thomas Kuerten, Tim Goodwin,
		Timothy L. Mayo, Todd Underwood, Tomoaki Terazawa,
		Tomoki Yoshioka, Toshinori Maeno, Uwe Ohse, Vern Hart,
		Vince Vielhaber, Waskita Adijarto, William E. Baxter,
		Wolfgang Rufeger, Wu Ching-hong, Yoshitatsu Takeshita.
20000307
	internal: switched to various new libraries.
	ui: tcpserver prints status on startup, before first connection.
20000309
	ui: incorporated rblsmtpd.
	ui: tcpserver -X allows nonexistent rules file.
	ui: http@ uses HTTP/1.0, sends Host, removes header.
	ui: added fixcrio.
	ui: delcr leaves CR alone at the end of a partial final line.
20000311
	ui: switched to prot; so setgid() is preceded by setgroups().
	ui: tcpserver supports -U.
	ui: tcpserver supports hostname rules.
	ui: tcprulescheck now uses environment variables.
20000312
	version: ucspi-tcp 0.86, beta.
20000314
	portability problem: the poll() emulation in RedHat 5.1 doesn't
		clear revents when select() returns 0. tnx Petr Novotny.
		impact: dns lookups busy-loop when they should time out.
		fix: new iopause from DNScache.
20000315
	version: ucspi-tcp 0.87, beta.
20000318
	internal: split rules() out of tcpserver and tcprulescheck.
	bug: didn't always clear rule length when $TCPREMOTEHOST was
		set. impact: empty rule could be ignored. fix: obvious.
		tnx Toshinori Maeno, Takashi Takizawa, Yuichi Katoh.
	ui: check for lone = rule if $TCPREMOTEHOST is set.
	ui: tcpclient supports fast+slow timeouts, 2+58 by default.
	ui: tcpclient randomizes order of addresses.
	version: ucspi-tcp 0.88, beta.
20080227 (feh)
	added Greetdelay option into rblsmtpd.

src/rblsmtpd.c
#include <unistd.h>
#include "byte.h"
#include "str.h"
#include "scan.h"
#include "fmt.h"
#include "env.h"
#include "exit.h"
#include "sig.h"
#include "buffer.h"
#include "getoptb.h"
#include "logmsg.h"
#include "stralloc.h"
#include "commands.h"
#include "pathexec.h"
#include "dnsresolv.h"
#include "ip.h"

#define WHO "rblsmtpd"

void nomem(void)
{
  logmsg(WHO,111,FATAL,"out of memory");
}
void usage(void)
{
  logmsg(WHO,100,USAGE,"rblsmtpd \
[ -B ] [ -b ] \
[ -C ] [ -c ] \
[ -i ] [ -t timeout ] \
[ -r base ] [ -a base ] \
[-W] [-w delay] smtpd [ arg ... ]");
}

char *ip_env = 0;;
static stralloc ip_reverse;
int flagip6;

void ip_init(void)
{
  unsigned int i;
  unsigned int j;
  char hexval;
  char remoteip[16];
 
  flagip6 = 0;
  byte_zero(remoteip,16);

  ip_env = env_get("TCP6REMOTEIP");
  if (ip_env) { 
    if (byte_equal(ip_env,7,V4MAPPREFIX)) 
      ip_env = ip_env + 7;
    else 
      flagip6 = 1;
  } else {
    ip_env  = env_get("TCPREMOTEIP");
    if (!ip_env) ip_env = "";
  }

  if (!stralloc_copys(&ip_reverse,"")) nomem();

  if (flagip6) { 
    if ((ip6_scan(ip_env,remoteip)) == 0) return;

    for (j = 15; j >= 0; j--) {
      hexval = tohex(remoteip[j] & 15);
      if (!stralloc_catb(&ip_reverse,&hexval,1)) nomem();
      if (!stralloc_cats(&ip_reverse,".")) nomem();

      hexval = tohex(remoteip[j] >> 4 & 15);
      if (!stralloc_catb(&ip_reverse,&hexval,1)) nomem();
      if (!stralloc_cats(&ip_reverse,".")) nomem(); 
      if (!j) break;
    }
  } 
  else {
    i = str_len(ip_env);
    while (i) {
      for (j = i; j > 0; --j) 
        if (ip_env[j - 1] == '.') break;
      if (!stralloc_catb(&ip_reverse,ip_env + j,i - j)) nomem();
      if (!stralloc_cats(&ip_reverse,".")) nomem();
      if (!j) break;
      i = j - 1;
    }
  }
}

unsigned long timeout = 60;
int flagrblbounce = 0;
int flagfailclosed = 0;
int flagmustnotbounce = 0;
int flagrblinfo = 0;

int decision = 0; /* 0 undecided, 1 accept, 2 direct refuse, 3 direct bounce, 4 rbl refuse, 5 rbl bounce */
static stralloc text; /* defined if decision is > 2 */

static stralloc tmp;

void rbl(char *base)
{
  if (decision) return;
  if (!stralloc_copy(&tmp,&ip_reverse)) nomem();
  if (!stralloc_cats(&tmp,base)) nomem();
  if (dns_txt(&text,&tmp) < 0) {
    flagmustnotbounce = 1;
    if (flagfailclosed) {
      if (!stralloc_copys(&text,"temporary RBL lookup error")) nomem();
      decision = 4;
    }
    return;
  }
  if (text.len) {
    if (flagrblbounce) 
      decision = 5;
    else 
      decision = 4;
  }
}

void antirbl(char *base)
{
  if (decision) return;
  int flagip;

  if (!stralloc_copy(&tmp,&ip_reverse)) nomem();
  if (!stralloc_cats(&tmp,base)) nomem();
   
  if (flagip6) 
    flagip = dns_ip6(&text,&tmp);
  else
    flagip = dns_ip4(&text,&tmp);
 
  if (flagip < 0) {
    flagmustnotbounce = 1;
    if (!flagfailclosed)
      decision = 1;
    return;
  }
  if (text.len)
    decision = 1;
}

char strnum[FMT_ULONG];
static stralloc message = {0};;
static stralloc info = {0};

char inspace[64]; 
buffer in = BUFFER_INIT(read,0,inspace,sizeof(inspace));
char outspace[1]; 
buffer out = BUFFER_INIT(write,1,outspace,sizeof(outspace));

void waitdelay(unsigned long delay)
{
  unsigned long u;
  char *x;

  if (!delay) {
    x = env_get("GREETDELAY");
    if (x) { 
      scan_ulong(x,&u); 
      delay = u; 
    }
  }

  if (delay) {
    if (!stralloc_copys(&info,"greetdelay: ")) nomem();

    buffer_puts(buffer_2,"rblsmtpd: ");
    buffer_puts(buffer_2,ip_env);
    buffer_puts(buffer_2," pid ");
    buffer_put(buffer_2,strnum,fmt_ulong(strnum,getpid()));
    buffer_puts(buffer_2,": ");
    buffer_put(buffer_2,info.s,info.len);
    buffer_put(buffer_2,strnum,fmt_ulong(strnum,delay));
    buffer_puts(buffer_2,"\n");
    buffer_flush(buffer_2);

    if (!stralloc_cats(&info,"\r\n")) nomem();

    sleep(delay);
  }
}

void rblinfo(void)
{
  int i; 

  if (!stralloc_copys(&message,"[RBL info] ")) nomem();

  if (text.len > 200) text.len = 200;
  if (!stralloc_cat(&message,&text)) nomem();

  for (i = 0; i < message.len; ++i)
    if ((message.s[i] < 32) || (message.s[i] > 126))
      message.s[i] = '?';

  buffer_puts(buffer_2,"rblsmtpd: ");
  buffer_puts(buffer_2,ip_env);
  buffer_puts(buffer_2," pid ");
  buffer_put(buffer_2,strnum,fmt_ulong(strnum,getpid()));
  buffer_puts(buffer_2,": ");
  buffer_put(buffer_2,message.s,message.len);
  buffer_puts(buffer_2,"\n");
  buffer_flush(buffer_2);

  if (!stralloc_copy(&info,&message)) nomem();
  if (!stralloc_0(&info)) nomem();
  if (!pathexec_env("RBLSMTPD",info.s)) nomem();
}

void reject() { buffer_putflush(&out,message.s,message.len); }
void accept() { buffer_putsflush(&out,"250 rblsmtpd.local\r\n"); }
void greet()  { buffer_putsflush(&out,"220 rblsmtpd.local\r\n"); }
void quit()   { buffer_putsflush(&out,"221 rblsmtpd.local\r\n"); _exit(0); }
void drop()   { _exit(0); }

struct commands smtpcommands[] = {
  { "quit", quit, 0 }
, { "helo", accept, 0 }
, { "ehlo", accept, 0 }
, { "mail", accept, 0 }
, { "rset", accept, 0 }
, { "noop", accept, 0 }
, { 0, reject, 0 }
} ;

void rblsmtpd(void)
{
  int i;

  if (flagmustnotbounce || (decision == 2) || (decision == 4)) {
    if (!stralloc_copys(&message,"451 ")) nomem();
  }
  else
    if (!stralloc_copys(&message,"553 ")) nomem();

  if (text.len > 200) text.len = 200;
  if (!stralloc_cat(&message,&text)) nomem();
  for (i = 0; i < message.len; ++i)
    if ((message.s[i] < 32) || (message.s[i] > 126))
      message.s[i] = '?';

  buffer_puts(buffer_2,"rblsmtpd: ");
  buffer_puts(buffer_2,ip_env);
  buffer_puts(buffer_2," pid ");
  buffer_put(buffer_2,strnum,fmt_ulong(strnum,getpid()));
  buffer_puts(buffer_2,": ");
  buffer_put(buffer_2,message.s,message.len);
  buffer_puts(buffer_2,"\n");
  buffer_flush(buffer_2);

  if (!stralloc_cats(&message,"\r\n")) nomem();

  if (!timeout)
    reject();
  else {
    sig_catch(sig_alarm,drop);
    alarm(timeout);
    greet();
    commands(&in,smtpcommands);
  }
  _exit(0);
}

int main(int argc,char **argv,char **envp)
{
  char *x;
  int opt;
  unsigned long greetdelay = 0;

  ip_init();

  x = env_get("RBLSMTPD");
  if (x) {
    if (!*x)
      decision = 1;
    else if (*x == '-') {
      if (!stralloc_copys(&text,x + 1)) nomem();
      decision = 3;
    }
    else {
      if (!stralloc_copys(&text,x)) nomem();
      decision = 2;
    }
  }

  while ((opt = getopt(argc,argv,"bBcCit:r:a:w:W")) != opteof)
    switch(opt) {
      case 'b': flagrblbounce = 1; break;
      case 'B': flagrblbounce = 0; break;
      case 'c': flagfailclosed = 1; break;
      case 'C': flagfailclosed = 0; break;
      case 'i': flagrblinfo = 1; break;
      case 't': scan_ulong(optarg,&timeout); break;
      case 'r': rbl(optarg); break;
      case 'a': antirbl(optarg); break;
      case 'W': if (!decision) { waitdelay(greetdelay); } break;
      case 'w': if (!decision) { scan_ulong(optarg,&greetdelay); waitdelay(greetdelay); } break;
      default: usage();
    }

  argv += optind;
  if (!*argv) usage();
 
  if (flagrblinfo && decision >= 4)
    rblinfo(); 
  else if (decision >= 2)
    rblsmtpd();

  pathexec(argv);
  logmsg(WHO,111,FATAL,B("unable to run: ",*argv));

  return 0;
}
